### ISSUE TITLE
Soft deleted models arg directive

### ISSUE THREAD

**Issue by olivernybroe** [nuwave/lighthouse#100]
Was thinking if we should add a directive for soft deleted models?

So we have an arg directive called `@trashes` or `@softdeletes` or something better than that, this arg directive would work in the following way.

``` 
type Query{
    companies(trashed: Trash @trashes): [Company!]! @paginate(model: "Company", type: "relay")
}

enum Trash {
  ONLY @enum(value: "only")
  WITH @enum(value: "with")
  WITHOUT @enum(value: "without")
}

{
  companies(first: 10, trashed: ONLY ) {
    edges {
      node {
        id,
        vat,
        name
      }
    }
  }
}
```

Of course we would store the Trash enum, so users don't have to define that themself.

**Comment by chrissm79** [comment]
@olivernybroe Interesting, and I wonder if we even need a directive here... we could just have an internally defined Enum that the `hasMany`, `paginate` directive would check for and if it exists it would add it to he auto-generated query. Added it to `v2.x` label!

**Comment by olivernybroe** [comment]
@chrissm79 Hmm the problem I see with that is the `hasMany` could end up being a little confusing with all the parameters. 
For example if I want to give the user the power to choose if they want soft deleted models as well, how do I then define what the key in which they define that they want deleted models is called. eg. in the example above I called it `trashed`. 

However I think the `hasMany` should have the opportunity to define soft deletes also, just as defaults  instead.

**Comment by chrissm79** [comment]
@olivernybroe The schema is almost exactly as you have it, just w/out the directive (rather than adding another argument to the `hasMany` directive). Like so:

```graphql
# Lighthouse's schema
enum Trash {
  ONLY @enum(value: "only")
  WITH @enum(value: "with")
  WITHOUT @enum(value: "without")
}


# User's schema
type Query{
    companies(trashed: Trash): [Company!]! @paginate(model: "Company", type: "relay")
}

{
  companies(first: 10, trashed: ONLY ) {
    edges {
      node {
        id,
        vat,
        name
      }
    }
  }
}
```

Then, using just the `FieldInfo` the resolver would check if the `Trash` enum was included in the arguments and pass it along to the query builder if so. This way you're not having to add an `Enum` and a `Directive` on the same field.

**Comment by olivernybroe** [comment]
@chrissm79 Hmm I like the idea about looking at the `FieldInfo` object instead, that's pretty awesome.

**Comment by gijsbeijer** [comment]
Hi all, 
I'm looking for the same functionality

I thought I'd make a new Directive looking like:

```
<?php

namespace Nuwave\Lighthouse\Schema\Directives;

use Nuwave\Lighthouse\Schema\Directives\BaseDirective;
use Nuwave\Lighthouse\Support\Contracts\ArgBuilderDirective;

class WithTrashedDirective extends BaseDirective implements ArgBuilderDirective
{

	public function name() : string
	{
		return 'withTrashed';
	}

	 public function handleBuilder($builder, $value)
    {
        return $builder->withTrashed();
    }
}
````
But this doesn't seem to do anything?

I thought the nicest way to use this was if we could use it as follows:
```user(id: ID @eq): User @withTrashed @find```
or
```raises: [Raise!]! @withTrashed @all(model: "App\\Raise")```

Does anybody know how to go about this, can't get this to work!

Thanks!

**Comment by olivernybroe** [comment]
@gijsbeijer What you are trying to do is not the same. We wanted to do it as part of your query, it seems like you just want to always return the trashed models also. 
I think your question belongs in the slack/spectrum chat instead, or as a new issue.

**Comment by gijsbeijer** [comment]
@olivernybroe sorry about that. Thinking of it, doing this via the query would be a lot nicer.

**Comment by spawnia** [comment]
@gijsbeijer as the name implies, `ArgBuilderDirectives` only work if they are applied to arguments.

What you are doing would require a new directive type, a way to hook into the builder through field directives.

**Comment by lorado** [comment]
any updates here? I currently rely on this feature...

**Comment by spawnia** [comment]
@olivernybroe thanks to @lorado this is now a part of Lighthouse and will be released with the coming `v4.3.0`
